Add txn_isolation reference (#2991)
tarantool/doc: doc/book/box/atomic/txn_mode_mvcc.rst

..  _txn_mode_mvcc-options:

Setting the transaction isolation level
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> ..  note::

Reviewer: The note mentions terms that haven't yet been introduced. Consider moving the note lower.

Author: Thanks, moved the note.

    all others (for example, ``replace``) with ``read-committed``.

The transaction manager has three options for the :ref:`transaction isolation level <transaction_model_levels>`:

*   ``best-effort`` (default)

*   ``read-committed``

*   ``read-confirmed``

Using ``best-effort`` as the default option allows MVCC to consider the actions of transactions
independently and determine the best :ref:`isolation level <transaction_model_levels>` for them.
It increases the probability of successful completion of the transaction and helps to avoid possible conflicts.

To set the default isolation level with the other option, for example, 
to ``read-committed``, use the following command:

..  code-block:: lua

    box.cfg{txn_isolation = 'read-committed'}

> You can also set an isolation level for specific transactions in their ``box.begin()`` calls:

Reviewer: ```suggestion
You can also set an isolation level for a specific transaction in its ``box.begin()`` call:
```

..  code-block:: lua

    box.begin({tnx_isolation = 'best-effort'})

> In this case, there is additionally the ``default`` option. It sets the transaction's isolation level
> to the one that is set in ``box.cfg``.

Reviewer: ```suggestion
In this case, you can also use the ``default`` option. It sets the transaction's isolation level
to the one set in ``box.cfg``.
```

..  note::

    You can set the isolation level in the net.box :ref:`stream:begin() <net_box-stream_begin>` method

>     and :ref:`IPROTO_BEGIN <box_protocol-begin>` binary protocol request.

Reviewer: I'm not sure that you can interfere with the IPROTO_BEGIN request per se. It is under the hood of `stream:begin()`.

Author: If I read the [binary protocol intro](https://www.tarantool.io/en/doc/latest/dev_guide/internals/box_protocol/) right, then yes, it's possible.

Choosing the better option depends on whether you have conflicts or not. 

> If you have many conflicts, you should set the different options or use 

Reviewer: ```suggestion
If you have many conflicts, you should set a different option or use 
```

The transaction manager has four options for the transaction isolation level that you can set in ``box-cfg``:

> *   ``default``

Author: @rindblack15 I took the liberty to remove periods according to the [list punctuation guidelines](https://www.tarantool.io/en/doc/latest/contributing/docs/style/#lists). Hope you don't mind :)

*   ``best-effort``

*   ``read-committed``

*   ``read-confirmed``

> 

Reviewer 2: Let's describe a little bit all the options here, not just the best-effort. Otherwise, it looks like you'll have to guess the rest or read about it on the previous pages.

Author: Thanks.
I've rephrased this in such a way that here we only list options without the explanation (it's available under the link before the list). The sentence about `best-effort` explains why it was chosen as a default, not how it works. So this all seems to me pretty consistent.

To set the default isolation level with the other option, for example, 
to ``read-committed``, use the following command:

..  code-block:: lua

    box.cfg{default_txn_isolation = 'read-committed'}
 

> 

Reviewer 2: It would be nice if we could learn from this section that the isolation levels can also be set via IPROTO.

Author: Thanks, added this.
